File: ruby/leetcode/linked_lists.rb
# frozen_string_literal: true
# TOPIC: Linked List LeetCode Problems | ruby linked_lists.rb

class ListNode
  attr_accessor :val, :next
  def initialize(val = 0, nxt = nil)
    @val = val
    @next = nxt
  end
end

# TODO 1: Reverse Linked List — LC #206
#   def reverse_list(head)
#     # Iterative: prev=nil, curr=head; while curr: next=curr.next; curr.next=prev; prev=curr; curr=next
#   end

# TODO 2: Merge Two Sorted Lists — LC #21
#   def merge_two_lists(list1, list2)
#     # Dummy node; compare heads, advance the smaller one
#   end

# TODO 3: Linked List Cycle — LC #141
#   def has_cycle(head)
#     # Floyd's: slow=head, fast=head; slow=slow.next, fast=fast.next.next; meet if cycle
#   end

# TODO 4: Remove Nth Node From End of List — LC #19
#   def remove_nth_from_end(head, n)
#     # Two pointers n apart; when fast reaches end, slow is at target
#   end

# TODO 5: Merge K Sorted Lists — LC #23
#   def merge_k_lists(lists)
#     # Use min-heap (priority queue); push heads; pop min, add next from that list
#   end

# TODO 6: Reorder List — LC #143
#   def reorder_list(head)
#     # Find middle (slow/fast), reverse second half, merge two halves alternating
#   end

File: php/leetcode/linked_lists.php
<?php
declare(strict_types=1);
// TOPIC: Linked List LeetCode Problems | php linked_lists.php

class ListNode {
    public function __construct(
        public int $val = 0,
        public ?ListNode $next = null,
    ) {}
}

// TODO 1: Reverse Linked List — LC #206
//   function reverseList(?ListNode $head): ?ListNode
//   // Iterative: $prev=null, $curr=$head; swap pointers as you advance

// TODO 2: Merge Two Sorted Lists — LC #21
//   function mergeTwoLists(?ListNode $list1, ?ListNode $list2): ?ListNode
//   // Dummy node; compare heads, advance smaller

// TODO 3: Linked List Cycle — LC #141
//   function hasCycle(?ListNode $head): bool
//   // Floyd's: slow=head, fast=head; if they meet → cycle

// TODO 4: Remove Nth Node From End of List — LC #19
//   function removeNthFromEnd(?ListNode $head, int $n): ?ListNode
//   // Two pointers n apart

// TODO 5: Merge K Sorted Lists — LC #23
//   function mergeKLists(array $lists): ?ListNode
//   // Min-heap using SplMinHeap

// TODO 6: Reorder List — LC #143
//   function reorderList(?ListNode $head): void
//   // Find middle, reverse second half, merge alternating
